indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
-                  }
-                }
-            );
-          }
-
-          existing.add(hdfsPath);
+        if (isSnapshot(jarFile)) {
+          addSnapshotJarToClassPath(jarFile, intermediateClassPath, fs, conf);
+        } else {
+          addJarToClassPath(jarFile, distributedClassPath, intermediateClassPath, fs, conf);
         }
+      }
+    }
+  }
 
-        DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
+  private static void addJarToClassPath(
+      File jarFile,
+      Path distributedClassPath,
+      Path intermediateClassPath,
+      FileSystem fs,
+      Configuration conf
+  )
+      throws IOException
+  {
+    // Non-snapshot jar files are uploaded to the shared classpath.
+    final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
+    if (!fs.exists(hdfsPath)) {
+      // Muliple jobs can try to upload the jar here,
+      // to avoid them from overwriting files, first upload to intermediateClassPath and then rename to the distributedClasspath.
+      final Path intermediateHdfsPath = new Path(intermediateClassPath, jarFile.getName());
+      uploadJar(jarFile, intermediateHdfsPath, fs);
+      try {
+        log.info("Renaming jar to path[%s]", hdfsPath);
+        fs.rename(intermediateHdfsPath, hdfsPath);
+        if(!fs.exists(hdfsPath)){
+          throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
+        }
+      }
+      catch (IOException e) {
+        // rename failed, possibly due to race condition. check if some other job has uploaded the jar file.
+        if (!fs.exists(hdfsPath)) {
+          log.error(e, "IOException while Renaming jar file");
+          throw e;
+        }
+      } finally {
+        if (fs.exists(intermediateHdfsPath)) {
+          fs.delete(intermediateHdfsPath, false);
+        }
       }
     }
+    DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
+  }
+
+  private static void addSnapshotJarToClassPath(
+      File jarFile,
+      Path intermediateClassPath,
+      FileSystem fs,
+      Configuration conf
+  ) throws IOException
+  {
+    // Snapshot jars are uploaded to non shared intermediate directory.
+    final Path hdfsPath = new Path(intermediateClassPath, jarFile.getName());
+
+    // existing is used to prevent uploading file multiple times in same run.
+    if (!existing.contains(hdfsPath)) {
+      uploadJar(jarFile, hdfsPath, fs);
+      existing.add(hdfsPath);
+    }
+    DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
+  }
+
+  private static void uploadJar(File jarFile, final Path path,final  FileSystem fs) throws IOException
+  {
+    log.info("Uploading jar to path[%s]", path);
+    ByteStreams.copy(
+        Files.newInputStreamSupplier(jarFile),
+        new OutputSupplier<OutputStream>()
+        {
+          @Override
+          public OutputStream getOutput() throws IOException
+          {
+            return fs.create(path);
```

Comment:
can you elaborate more on which transient error you are referring to ? 
At present, this will overwrite any partial written file from previous run if any.